### PR TITLE
sandbox/apparmor: aare exclusion rule generation

### PR DESCRIPTION
**Pull request by jslarraz**
Followup of PR#11567 
- Core function `generateAAREExclusionPatternsGenericImpl` has been reworked to simplify it as much as possible 
- Tests have been updated to compare the compiled profile (generated by apparmor_parser) instead of profile strings

### CODE REVIEW

**Comment by alexmurray** [comment]
@jslarraz unit tests appear to be failing:

```
FAIL: docker_support_test.go:291: DockerSupportInterfaceSuite.TestGenerateAAREExclusionPatterns
```

**Review by alexmurray — APPROVED** [review]
LGTM - thanks for reworking this to be simpler and more easy to understand - and for the additional tests!

**Comment by alexmurray** [comment]
@zyga hey, any chance you or someone else from the snapd team can take a look at this one? Thanks!

File: interfaces/builtin/docker_support.go
Comment (pedronis) [inline interfaces/builtin/docker_support.go:201]: using %s feels fragile, my preference would be to use a placeholder that contains some hint of what would be produced or the arguments passed to the helper:

###EXCL{change_profile unsafe <>  -> docker-default,:/snap/snapd/*/usr/lib/snapd/snap-confine,/snap/core/*/usr/lib/snapd/snap-confine}###

so the strawman for the placeholder here would be: ###EXCL{$PREFIX<>$SUFFIX:$STRINGS}### but anything sensible could work

then you could have a helper that takes the string with the placeholders plus the arguments to generate things, and the helper computes both the placeholder from the arguments and the generated content, and does the replacement or errors if it doesn't find the placeholder at all

**Review by pedronis — COMMENTED** [review]
I find the use of %s a bit fragile, made a suggestion

**Comment by jslarraz** [comment]
Updated to use place holders as suggested

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:243]: I wonder if we should exclude the commonPrefix already here?

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:259]: instead of checking for len == 1 here, we could share more code and check for len(aux) below, as anyway len == 1 is not the common case?

**Review by pedronis — APPROVED** [review]
thanks for the changes, couple of small comments

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:154]: This should be `errors.New`

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:185]: Stray newline

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:75]: Stray newline

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:210]: I don't think there's a need for this to be on the heap.

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:212]: Stray newline.

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:266]: Do we need to format a string to add it to the builder? We can just
```go
builder.WriteString(opts.Prefix)
builder.WriteString(commonPrefix)
builder.WriteRune('{')
for i := range aux {
  if i > 0 {
    builder.WriteRune(',')
  }
  builder.WriteString(aux[i])
}
builder.WriteRune('}')
builder.WriteString(suffix)
```

This will be less wasteful on the Gc.

File: testutil/apparmor.go
Comment (zyga) [inline testutil/apparmor.go:10]: Stray newline again?

File: testutil/apparmor.go
Comment (zyga) [reply testutil/apparmor.go:31]: We can use `io.Copy` to efficiently read data from the process pipe and write it to the hasher. I think this should be improved. There's far less code to write without running `sha1sum` outside of go.

File: sandbox/apparmor/apparmor.go
Comment (zyga) [inline sandbox/apparmor/apparmor.go:194]: Please break this out to a helper function. It will make the rest more readable and shorter. I think it can go to the `strutil` package. Write a separate test for this please.

**Review by zyga — COMMENTED** [review]
Quick pass. I think it's nice but I would like to see some improvements before this lands.

Separately: can we agree to move away from ###STUFF### replacements and move to `text/template` which is safer and more efficient. This is a follow-up for this but I just want say we should stop pushing down that path where we had serious release-breaking bugs because we forgot to do string replacement right and it was slient.

I will do a more thorough review tomorrow.

File: interfaces/builtin/docker_support.go
Comment (bboozzoo) [inline interfaces/builtin/docker_support.go:201]: The patters, in not expanded, would be treated as valid comments and also their format is non trivial. We have a check for ensuring that a call to InsertAARE.. fails when the placeholder was not found, but we have no checks for the opposite case, i.e. someone adds a pattern, but there is no expansion for it. 

Now, we seem to have the assumptions that placeholders generally follow a pattern of `###.*###`, so I'm wondering if adding such check somewhere in Backend.addContent where we deal with `###SNIPPETS###` could be useful.

Maybe a followup? This PR is already quite large.

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:197]: hm the common prefix search only makes sense up to the length of the shortest pattern, doesn't it?
Also we already established that the first characters is the same in all patterns (it's `/`')

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:199]: and then the `len()` check isn't needed

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:212]:
```suggestion
```

**Comment by pedronis** [comment]
> Separately: can we agree to move away from ###STUFF### replacements and move to text/template which is safer and more efficient. This is a follow-up for this but I just want say we should stop pushing down that path where we had serious release-breaking bugs because we forgot to do string replacement right and it was slient.

that's a disussion for snapd,  the PR is just following what most things are doing in interfaces, which is placeholders/replacement based, that's an old decision and I wasn't actually part of it. There are probably challenges to use templates as well that need to be figure out on their own.

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:122]: out of curiosity, is it somehow easier for the apparmor to construct the DFA if rules are expressed such that we have a common prefix followed by `{}` expansion of alternatives?

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:211]: since we have the common prefix now, should this start from the rune right after the common prefix then?

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [inline sandbox/apparmor/apparmor.go:203]: I think we should check if the common prefix isn't by any chance equal to one of the patterns, i.e. a case say `/foo` and `/foo/bar`.

File: testutil/apparmor.go
Comment (bboozzoo) [inline testutil/apparmor.go:10]:
```suggestion
```

**Review by bboozzoo — COMMENTED** [review]
Looks nice. I left some comments, please have a look.

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: testutil/apparmor.go
Comment (jslarraz) [reply testutil/apparmor.go]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:77]: Done in commit 25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: `errors.New` does not accept format string, should it be `fmt.Errorf`? If so, done in commit https://github.com/snapcore/snapd/commit/25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit https://github.com/snapcore/snapd/commit/25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit https://github.com/snapcore/snapd/commit/25f75be5889fc9b06f51856734f4fb712dfe42fa

File: testutil/apparmor.go
Comment (jslarraz) [reply testutil/apparmor.go]: Done in commit https://github.com/snapcore/snapd/commit/25f75be5889fc9b06f51856734f4fb712dfe42fa

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:243]: For iterations of the outer loop `for charInd := 1; charInd < size; charInd++ ` where `charInd < len(commonPrefix)`, len(subpatternPrefix) will be smaller than len(commonPrefix). 

We could move the apparmor rules generation for the common prefix out of the main loop (`for charInd := 1; charInd < size; charInd++ `), however it will require to duplicate some checks (suffixes, '*' and '*/').

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:259]: It will be 1, at least for all the iterations of the outer loop `for charInd := 1; charInd < size; charInd++` where `charInd < len(commonPrefix)`. 

As in the previous comment, we could move the apparmor rules generation for the common prefix out of the main loop (for charInd := 1; charInd < size; charInd++ ), however it will require to duplicate some checks (suffixes, '' and '/').

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:211]: If I understand it correctly you mean the builder object itself, and not the builder content (string), which will be allocated in the heap anyway. So the suggestion is to use `builder := strings.Builder{}` instead. Is that right?

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:266]: It looks less readable and harder to see what is actually doing tbh. Do you think it would be worth the performance gain?

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:212]: Even if we have the common prefix, we still need to write the apparmor rules for the common prefix, what is done by 

```
		} else if len(subpatterns) == 1 {
			res := fmt.Sprintf("%s%s[^%s]**%s\n", opts.Prefix, subpatternPrefix, subpatterns[subpatternPrefix], opts.Suffix)
			builder.WriteString(res)
```

As in the previous comments, we could move the apparmor rules generation for the common prefix out of the main loop (`for charInd := 1; charInd < size; charInd++`), however it will require to duplicate some checks (suffixes, '*' and '*/').

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [reply sandbox/apparmor/apparmor.go]: sorry, yes I meant fmt.Errorf()

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:195]: Done in commit commit c7b920189797ba12ab53f47a5e3b8da1e01f162e

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Refactored to move the functionality to find the common prefix to strutil as suggested by @zyga. Done in commit commit c7b920189797ba12ab53f47a5e3b8da1e01f162e

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in commit commit c7b920189797ba12ab53f47a5e3b8da1e01f162e

**Comment by jslarraz** [comment]
Hey @pedronis @zyga @bboozzoo 

I did a first pass on your commentaries. Waiting for feedback.

File: sandbox/apparmor/apparmor.go
Comment (zyga) [reply sandbox/apparmor/apparmor.go:257]: I think that is the whole point of using the builder.

Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:257]: Umm, I'm now wondering whether we need a builder at all. I retained this pattern from the original PR but maybe simple string concatenation is enough for this one. What do you think?

**Comment by jslarraz** [comment]
Hey @zyga @bboozzoo, 

Any update so far? specially regarding https://github.com/snapcore/snapd/pull/13488#discussion_r1587274327

**Comment by zyga** [comment]
> Hey @zyga @bboozzoo, 
> 
> 
> 
> Any update so far? specially regarding https://github.com/snapcore/snapd/pull/13488#discussion_r1587274327

I'll review this soon

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [reply sandbox/apparmor/apparmor.go:257]: TBH builder is preferred to any form of string concatenation.

Comment (bboozzoo) [reply sandbox/apparmor/apparmor.go:257]: I think

File: strutil/commonprefix.go
Comment (bboozzoo) [inline strutil/commonprefix.go:25]:
```suggestion
```

File: sandbox/apparmor/apparmor.go
Comment (bboozzoo) [reply sandbox/apparmor/apparmor.go:202]: or
```
var builder strings.Builder
```

**Review by bboozzoo — COMMENTED** [review]
Looks good, please see the comment about using `strings.Builder` and then it's ready to land.

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go:257]: Fine for me, will do :)

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in https://github.com/snapcore/snapd/pull/13488/commits/5d3279b19f6c81d917f386a1d3985abccb317ff7

File: sandbox/apparmor/apparmor.go
Comment (jslarraz) [reply sandbox/apparmor/apparmor.go]: Done in https://github.com/snapcore/snapd/pull/13488/commits/5d3279b19f6c81d917f386a1d3985abccb317ff7

File: strutil/commonprefix.go
Comment (jslarraz) [reply strutil/commonprefix.go]: Done in https://github.com/snapcore/snapd/pull/13488/commits/d3da3cbe2adc3e39947df9fb9590e9afce87426a

**Comment by jslarraz** [comment]
Updated to use string builder in https://github.com/snapcore/snapd/pull/13488/commits/5d3279b19f6c81d917f386a1d3985abccb317ff7. Master merged and i/b/docker_support_tests updated to add components to interfaces.SnapAppSet

**Comment by bboozzoo** [comment]
cross distro unit tests are failing, please have a look. I suspect it's likely due to SnapMountDir being different on Fedora/openSUSE. You'll likely need to call `release.MockOSRelease(..)`

**Comment by jslarraz** [comment]
Tests in sandbox/apparmor/apparmor_test.go were designed to compile the apparmor profile under tests using apparmor_parser and compare compiled version against the expected. When apparmor_parser is not available (as it seems to be the case in Fedora/openSUSE) test falls back to raw string comparison of the profiles. That caused some issues I addressed in https://github.com/snapcore/snapd/pull/13488/commits/e43153dc086140704d7392ebe407a37e9e34b786 and https://github.com/snapcore/snapd/pull/13488/commits/d064ae5b35fec8907d222c046b2e7616d0f2d362. 

For interfaces/builtin/docker_support_test.go, falling back to string comparison is not desirable as it will require to modify the current version of the apparmor rules to fit the exact string generated by GenerateAAREExclusionPatterns, loosing its intention to act as a regression test. Maybe disabling this tests if apparmor_parser is not available is a better approach here. Thoughts? 

```
docker_support_test.go:859:
    c.Assert(err, IsNil)
... value *exec.Error = &exec.Error{Name:"apparmor_parser", Err:(*errors.errorString)(0x1242140)} ("exec: \"apparmor_parser\": executable file not found in $PATH")
```

**Comment by bboozzoo** [comment]
> Tests in sandbox/apparmor/apparmor_test.go were designed to compile the apparmor profile under tests using apparmor_parser and compare compiled version against the expected. When apparmor_parser is not available (as it seems to be the case in Fedora/openSUSE) test falls back to raw string comparison of the profiles. That caused some issues I addressed in [e43153d](https://github.com/snapcore/snapd/commit/e43153dc086140704d7392ebe407a37e9e34b786) and [d064ae5](https://github.com/snapcore/snapd/commit/d064ae5b35fec8907d222c046b2e7616d0f2d362).
> 
> For interfaces/builtin/docker_support_test.go, falling back to string comparison is not desirable as it will require to modify the current version of the apparmor rules to fit the exact string generated by GenerateAAREExclusionPatterns, loosing its intention to act as a regression test. Maybe disabling this tests if apparmor_parser is not available is a better approach here. Thoughts?
> 
> ```
> docker_support_test.go:859:
>     c.Assert(err, IsNil)
> ... value *exec.Error = &exec.Error{Name:"apparmor_parser", Err:(*errors.errorString)(0x1242140)} ("exec: \"apparmor_parser\": executable file not found in $PATH")
> ```

Yes, skipping the test sounds fine if the parser isn't available. I think I should ultimately fix the unit test and install runtime dependencies as well, but it's a lower priority, so feel free to simply call `c.Skip()` when the parser is missing.

**Comment by jslarraz** [comment]
@bboozzoo Updated to skip the test when apparmorparser is not usable

**Comment by ernestl** [comment]
Not critical for 2.64, but nice to have

**Comment by ernestl** [comment]
@zyga, please have another look

**Review by bboozzoo — APPROVED** [review]
Thanks

**Comment by bboozzoo** [comment]
@jslarraz can you rebase once more?

**Pull request merged**
